### DISCUSSION TITLE
Login input detection

### DISCUSSION THREAD

**Discussion by jonaaix** [RoelVB/ChromeKeePass#35] [General]
I actually didn't find yet how the login input field detection works. I would be open to contribute some improvements.
There are a couple of sites, where the login field is not recognized, especially if it is behind a modal.

- [ ] [instagram.com](https://www.instagram.com/accounts/login/)
- [ ] [twitter.com](https://twitter.com/account/add)
- [ ] [soundcloud.com](https://soundcloud.com)
- [ ] [songsterr.com](https://www.songsterr.com/a/wa/signin)

You can comment more, if you have the issue.

**Comment by RoelVB** [comment]
Sometime you have to "redetect fields" using the right-click menu or the keyboardshortcut.
I've tested this with Instagram and Twitter and it works fine.

I appreciate you wanting to contribute. This is how the detection flow is now:
The `content_script` calls the `detectFields()` method after the page is loaded:
https://github.com/RoelVB/ChromeKeePass/blob/6826fdf9d2d914b6a2a3b9dce179720b7f7e9a43/src/content_script.ts#L6-L8

Then the `PageControl` class tries to find credential fields:
https://github.com/RoelVB/ChromeKeePass/blob/6826fdf9d2d914b6a2a3b9dce179720b7f7e9a43/src/classes/PageControl.ts#L27-L67

**Comment by jonaaix** [comment]
https://www.tchibo.de/bettwaesche-grau-p402011966.html#modal-productimagegallery-modalGalleryImage-401449333

The input field has attribute `autocomplete="off"` but ChromeKeepass detects it as login input

**Reply by RoelVB** [reply]
A field with `autocomplete="off"` could be a login field, so I'm not sure if we should ignore this.
Maybe make it optional in next version?

**Reply by jonaaix** [reply]
I'm currently collecting all issues I'm running into, to be sure to have most edge cases covered when improving the detection. I could also imagine to use machine learning technologies here to accurately detect the correct input fields.

**Comment by RoelVB** [comment]
Twitter and Instagram should work fine in the next update.
Because of a24141f316047decf27dd690472e9a7f092bf5c5

**Reply by jonaaix** [reply]
Probably we can use a MutationOberserver instead of interval checks. Seems to be widely supported so far. But I think it will require some performance tests.

https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver
https://caniuse.com/mutationobserver

**Reply by RoelVB** [reply]
That's a great idea. Luckily it's already in the `dev` branch (it's going to be in the next release) 😉 
I did some performance tests and it has literally no CPU load worth mentioning, but there will still be an option to disable it.

I tested on a couple af sites with livechats, because they update all the time. But the extensions CPU load stayed at 0.0%.

**Reply by jonaaix** [reply]
Sounds great!
I think where it could create issues are sites, that have a lot of DOM interaction because they are rendered via JavaScript using Angular, React or OpenGL. You could also try those sites:

- https://threejs.org/examples/#webgl_instancing_performance
- https://material.angular.io/components/categories
- Facebook
- Airbnb

**Comment by jonaaix** [comment]
![image](https://user-images.githubusercontent.com/15804690/114280956-f7701d80-9a3b-11eb-84ba-33761eddbb76.png)

Unfortunately the plugin recognized this field as login input and filled in the values, but I was editing a wordpress user profile, and the prefilled values got saved.

Probably we can disallow larger forms in the detection, and I would not use `autocomplete="off"` as single detection criterion. Or alternatively if it is the single criterion, then disallow the prefill.

![image](https://user-images.githubusercontent.com/15804690/114281056-7e24fa80-9a3c-11eb-8b2f-76b23e1676b6.png)

**Reply by RoelVB** [reply]
It's going to be hard to detect "larger forms", because a lot of modern sites don't use `<form>`-tags anymore. So what defines a form?
In your example we could say the table contains a few input fields, which could make it a "larger form", but this could break detection on other sites where there's a lot of/all content in a table. It's hard to imagine now, but there are still some older sites around that are build entirely with tables in tables, next to tables. Because bang in the day is was the easiest way to get your layout the same in all the different browsers.

Is it really a problem that the field is detected incorrectly? I guess it's problematic when you have "automatically fill single credentials" enabled.

But... I would like a future version to have an option to blacklist/whitelist sites or even fields on sites.

**Reply by jonaaix** [reply]
Currently I think the most convenient way would be to keep the fact, that most sites would use the correct input types `type="email"` and `type="password"` and probably have some login related wording (`.*pass.*`, `.*user.*`, `.*mail.*`, ...) in the input ids. But I see this is very difficult to predict. So having a validated growing dataset with columns of attribute names and values, tree distance to the corresponding other login inputs, would help in finding a pattern. I'm pretty sure that for each site there can be multiple different conditions fulfilled for login inputs. Then we can use a threshold of n conditions that must be fulfilled, and use different quantifier for the conditions (necessary and sufficient).
And finally I think there will always be some sites, that will need a manual selection. Some AdBlockers also use this method of defining explicit elements.
